string_constantt::to_array_expr must maintain the source location

tautschnig opened the pull request:
The C front-end uses this method to create arrays from string literals.
The resulting expression must have a source location to ensure the
resulting GOTO instruction knows which source location to use.

Fixes: #6994

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
string_constantt::to_array_expr must maintain the source location

The C front-end uses this method to create arrays from string literals.
The resulting expression must have a source location to ensure the
resulting GOTO instruction knows which source location to use.

Fixes: #6994

tautschnig committed:
Remove declaration of string_constantt::from_array_expr

The implementation was removed in a9043003de.

kroening commented on src/util/string_constant.cpp:
```diff
@@ -53,5 +53,8 @@ array_exprt string_constantt::to_array_expr() const
 
+  if(source_location().is_not_nil())
+    dest.add_source_location() = source_location();
+
```
We might like to have a `copy_source_location(...)` method or something similar.

kroening replied:
And maybe a "fluent" variant, say named `with_source_location(...)`

tautschnig replied:
I have only done the fluent interface for now.

tautschnig committed:
exprt::with_source_location: conditionally copy source location

This helper will copy the source location from another expression, if it
has any. Implemented as a fluent interface, with a first specialisation
for array_exprt. Other expression types an use
with_source_location<X_exprt> or add specialisations.
